calico_containers/pycalico/block.py

```diff
@@ -0,0 +1,383 @@
+        :return: None.
+        """
+        assert isinstance(address, IPAddress)
+        # Convert to an ordinal
+        ordinal = int(address - self.cidr.first)
+        assert 0 <= ordinal <= BLOCK_SIZE, "Address not in block."
+
+        # Check if allocated
+        if self.allocations[ordinal] is not None:
+            raise AlreadyAssignedError("%s is already assigned in block %s" % (
+                address, self.cidr))
+
+        # Set up attributes
+        attr_index = self._find_or_add_attrs(primary_key, attributes)
+        self.allocations[ordinal] = attr_index
+        return
+
+    def count_free_addresses(self):
+        """
+        Count the number of free addresses in this block.
+        :return: Number of free addresses.
+        """
+        count = 0
+        for a in self.allocations:
+            if a is None:
+                count += 1
+        return count
+
+    def release(self, addresses):
+        """
+        Release the given addresses.
+
+        :param addresses: Set of IPAddresses to release.
+        :return: Set of IPAddresses.  If any of the requested addresses were
+        not allocated, they are returned so the caller can handle
+        appropriately.
+        """
+        assert isinstance(addresses, (set, frozenset))
+        deleting_ref_counts = {}
+        ordinals = []
+        unallocated = set()
+        for address in addresses:
+            assert isinstance(address, IPAddress)
+            # Convert to an ordinal
+            ordinal = int(address - self.cidr.first)
+            assert 0 <= ordinal <= BLOCK_SIZE, "Address not in block."
+
+            # Check if allocated
+            attr_idx = self.allocations[ordinal]
+            if attr_idx is None:
+                _log.warning("Asked to release %s in block %s, but it was not "
+                             "allocated.", address, self.cidr)
+                unallocated.add(address)
+                continue
+            ordinals.append(ordinal)
+            old_count = deleting_ref_counts.get(attr_idx, 0)
+            deleting_ref_counts[attr_idx] = old_count + 1
+
+        # Compute which attributes need to be cleaned up.  We do this by
+        # reference counting.  If we're deleting all the references, then it
+        # needs to be cleaned up.
+        attr_index_to_delete = set()
+        ref_counts = self._get_attribute_ref_counts()
+        for idx, refs in deleting_ref_counts.iteritems():
+            if ref_counts[idx] == refs:
+                attr_index_to_delete.add(idx)
+
+        # Delete attributes if necessary
+        if attr_index_to_delete:
+            self._delete_attributes(attr_index_to_delete, ordinals)
+
+        # All attributes updated.  Finally, release all the requested
+        # addressses.
+        for ordinal in ordinals:
+            self.allocations[ordinal] = None
+
+        return unallocated
+
+    def _delete_attributes(self, attr_index_to_delete, ordinals):
+        """
```

Comment:
Worth saying that this rebalances (not really the correct word) the attribute list and indexes.